File: packages/cli/src/commands/review.ts
/**
 * `matimo review` — Human oversight for agent-created tools.
 *
 * Subcommands:
 *   matimo review list              Show all tools awaiting approval
 *   matimo review approve <name>    Approve a pending tool (requires MATIMO_APPROVAL_SECRET)
 *   matimo review reject  <name>    Reject / revoke a tool's approval
 */

import path from 'path';
import * as YAML from 'yaml';

/** Subset of ApprovalManifest we use here (avoids stale-dist type issues). */
interface ManifestHandle {
  getPendingTools(): string[];
  listApproved(): string[];
  getApproval(name: string): { approvedAt: string; approvedBy?: string } | undefined;
  revoke(name: string): boolean;
  approve(name: string, hash: string, approvedBy?: string): void;
  computeHash(content: string): string;
}

function resolveManifestDir(): string {
  // Look for the approvals file in cwd or the MATIMO_TOOL_DIR env var.
  const toolDir = process.env.MATIMO_TOOL_DIR;
  return toolDir ? path.resolve(toolDir) : process.cwd();
}

/** Try to import @matimo/core and return ApprovalManifest, or null if unavailable. */
async function tryLoadManifest(dir: string): Promise<ManifestHandle | null> {
  try {
    // Dynamic import works in both ESM (built CLI) and CJS (ts-jest test transform)
    const core = (await import('@matimo/core')) as Record<string, unknown>;
    const ApprovalManifest = core['ApprovalManifest'] as new (dir: string) => ManifestHandle;
    return new ApprovalManifest(dir);
  } catch {
    return null;
  }
}

function printTable(headers: string[], rows: string[][]): void {
  const cols = headers.length;
  const widths = headers.map((h, i) => Math.max(h.length, ...rows.map((r) => (r[i] ?? '').length)));
  const divider = widths.map((w) => '─'.repeat(w + 2)).join('┼');
  const fmt = (row: string[]) =>
    '│ ' + row.map((cell, idx) => cell.padEnd(widths[idx])).join(' │ ') + ' │';

  console.info('┌' + widths.map((w) => '─'.repeat(w + 2)).join('┬') + '┐');
  console.info(fmt(headers));
  console.info('├' + divider + '┤');
  for (const row of rows) {
    console.info(fmt(row.map((c) => c ?? '').slice(0, cols)));
  }
  console.info('└' + widths.map((w) => '─'.repeat(w + 2)).join('┴') + '┘');
}

async function listPending(dir: string): Promise<void> {
  const manifest = await tryLoadManifest(dir);
  if (!manifest) {
    console.error('❌ @matimo/core is not available. Run `pnpm install` first.');
    process.exit(1);
  }

  const pending = manifest.getPendingTools();
  const approved = manifest.listApproved();

  if (pending.length === 0 && approved.length === 0) {
    console.info('ℹ️  No tools are pending or approved.');
    return;
  }

  if (pending.length > 0) {
    console.info('\n⏳ Pending approval:\n');
    const rows = pending.map((name: string) => [name, 'pending', '—', '—']);
    printTable(['Tool name', 'Status', 'Approved by', 'Approved at'], rows);
  }

  if (approved.length > 0) {
    console.info('\n✅ Approved tools:\n');
    const rows = approved.map((name) => {
      const rec = manifest.getApproval(name)!;
      return [name, 'approved', rec.approvedBy ?? '—', rec.approvedAt];
    });
    printTable(['Tool name', 'Status', 'Approved by', 'Approved at'], rows);
  }

  if (pending.length > 0) {
    console.info(
      `\nRun "matimo review approve <tool-name>" to approve, or "matimo review reject <tool-name>" to reject.`
    );
  }
}

async function approveTool(toolName: string, dir: string): Promise<void> {
  if (!toolName) {
    console.error('❌ Usage: matimo review approve <tool-name>');
    process.exit(1);
  }

  const manifest = await tryLoadManifest(dir);
  if (!manifest) {
    console.error('❌ @matimo/core is not available. Run `pnpm install` first.');
    process.exit(1);
  }

  const pending = manifest.getPendingTools();
  if (!pending.includes(toolName)) {
    const approved = manifest.listApproved();
    if (approved.includes(toolName)) {
      console.info(`ℹ️  "${toolName}" is already approved.`);
      return;
    }
    console.error(
      `❌ No pending tool named "${toolName}". Run "matimo review list" to see pending tools.`
    );
    process.exit(1);
  }

  // Require a human-set approval secret — not the auto-generated one
  const secret = process.env.MATIMO_APPROVAL_SECRET;
  if (!secret) {
    console.error(
      '❌ MATIMO_APPROVAL_SECRET is not set.\n' +
        '   Set it to approve tools: export MATIMO_APPROVAL_SECRET=<your-secret>'
    );
    process.exit(1);
  }

  // Find the definition YAML to hash it
  const fs = await import('fs');
  const toolDir = path.join(dir, toolName);
  const yamlPath = path.join(toolDir, 'definition.yaml');
  if (!fs.existsSync(yamlPath)) {
    console.error(`❌ Cannot find definition.yaml for tool "${toolName}" at:\n   ${yamlPath}`);
    process.exit(1);
  }

  const originalYamlContent = fs.readFileSync(yamlPath, 'utf-8');
  let finalYamlContent = originalYamlContent;

  // Promote status in definition.yaml to "approved" so runtime policy sees the approval.
  try {
    const parsed = YAML.parse(originalYamlContent) ?? {};
    if (parsed && typeof parsed === 'object') {
      const currentStatus = (parsed as { status?: string }).status;
      if (currentStatus !== 'approved') {
        (parsed as { status?: string }).status = 'approved';
        finalYamlContent = YAML.stringify(parsed);
        // Write atomically using tmp + rename pattern
        const tmpPath = yamlPath + '.tmp';
        fs.writeFileSync(tmpPath, finalYamlContent, 'utf-8');
        fs.renameSync(tmpPath, yamlPath);
        console.info(`   📝 Updated status: draft → approved in definition.yaml`);
      }
    }
  } catch {
    console.warn(
      '⚠️  Failed to update status in definition.yaml; proceeding with manifest approval only.'
    );
  }

  const hash = manifest.computeHash(finalYamlContent);
  const approvedBy = process.env.USER ?? process.env.USERNAME ?? 'cli';
  manifest.approve(toolName, hash, approvedBy);

  console.info(`✅ Tool "${toolName}" approved by ${approvedBy}.`);
}

async function rejectTool(toolName: string, dir: string): Promise<void> {
  if (!toolName) {
    console.error('❌ Usage: matimo review reject <tool-name>');
    process.exit(1);
  }

  const manifest = await tryLoadManifest(dir);
  if (!manifest) {
    console.error('❌ @matimo/core is not available. Run `pnpm install` first.');
    process.exit(1);
  }

  const wasApproved = manifest.revoke(toolName);
  // Remove from pending even if it wasn't in approved set
  // (revoke only removes from approved; pending is cleaned up on next loadFromDisk)
  // Re-save by marking and un-marking is not clean — just confirm action
  const pending = manifest.getPendingTools();
  const wasPending = pending.includes(toolName);

  if (!wasApproved && !wasPending) {
    console.info(`ℹ️  No record of tool "${toolName}". Nothing to reject.`);
    return;
  }

  console.info(`🗑  Tool "${toolName}" has been rejected/revoked.`);
  if (wasApproved) {
    console.info('   (Approval signature removed — the tool will be blocked until re-approved.)');
  }
}

export async function reviewCommand(args: string[]): Promise<void> {
  const sub = args[0];
  const dir = resolveManifestDir();

  switch (sub) {
    case 'list':
    case undefined:
      await listPending(dir);
      break;
    case 'approve':
      await approveTool(args[1], dir);
      break;
    case 'reject':
      await rejectTool(args[1], dir);
      break;
    default:
      console.error(`❌ Unknown review subcommand: "${sub}"`);
      console.info('Usage: matimo review [list|approve|reject] [tool-name]');
      process.exit(1);
  }
}

File: python/packages/cli/src/matimo_cli/commands/review.py
"""
``matimo review`` — human oversight for agent-created tools.

Mirrors: packages/cli/src/commands/review.ts

Subcommands:
  matimo review list              Show all tools awaiting approval
  matimo review approve <name>    Approve a pending tool
  matimo review reject  <name>    Reject / revoke a tool
"""
from __future__ import annotations

import os
import sys
from pathlib import Path

import yaml


def _resolve_manifest_dir() -> str:
    tool_dir = os.environ.get("MATIMO_TOOL_DIR")
    return str(Path(tool_dir).resolve()) if tool_dir else os.getcwd()


def _try_load_manifest(directory: str):
    """Try to import ApprovalManifest from matimo core."""
    try:
        from matimo.policy.approval_manifest import ApprovalManifest  # type: ignore[import-not-found]
        return ApprovalManifest(directory)
    except ImportError:
        return None


def _print_table(headers: list[str], rows: list[list[str]]) -> None:
    cols = len(headers)
    widths = [
        max(len(h), *(len(r[i]) if i < len(r) else 0 for r in rows))
        for i, h in enumerate(headers)
    ]
    fmt = lambda row: "│ " + " │ ".join(c.ljust(w) for c, w in zip(row, widths)) + " │"

    print("┌" + "┬".join("─" * (w + 2) for w in widths) + "┐")
    print(fmt(headers))
    print("├" + "┼".join("─" * (w + 2) for w in widths) + "┤")
    for row in rows:
        padded = (row + [""] * cols)[:cols]
        print(fmt(padded))
    print("└" + "┴".join("─" * (w + 2) for w in widths) + "┘")


def _list_pending(directory: str) -> None:
    manifest = _try_load_manifest(directory)
    if manifest is None:
        print("❌ matimo core is not available. Run `pip install matimo` first.", file=sys.stderr)
        sys.exit(1)

    pending = manifest.get_pending_tools()
    approved = manifest.list_approved()

    if not pending and not approved:
        print("ℹ️  No tools are pending or approved.")
        return

    if pending:
        print("\n⏳ Pending approval:\n")
        rows = [[name, "pending", "—", "—"] for name in pending]
        _print_table(["Tool name", "Status", "Approved by", "Approved at"], rows)

    if approved:
        print("\n✅ Approved tools:\n")
        rows = []
        for name in approved:
            rec = manifest.get_approval(name) or {}
            rows.append([name, "approved", rec.get("approved_by", "—"), rec.get("approved_at", "—")])
        _print_table(["Tool name", "Status", "Approved by", "Approved at"], rows)

    if pending:
        print('\nRun "matimo review approve <tool-name>" to approve, '
              'or "matimo review reject <tool-name>" to reject.')


def _approve_tool(tool_name: str, directory: str) -> None:
    if not tool_name:
        print("❌ Usage: matimo review approve <tool-name>", file=sys.stderr)
        sys.exit(1)

    manifest = _try_load_manifest(directory)
    if manifest is None:
        print("❌ matimo core is not available.", file=sys.stderr)
        sys.exit(1)

    pending = manifest.get_pending_tools()
    if tool_name not in pending:
        approved = manifest.list_approved()
        if tool_name in approved:
            print(f'ℹ️  "{tool_name}" is already approved.')
            return
        print(f'❌ No pending tool named "{tool_name}". Run "matimo review list" to see pending tools.', file=sys.stderr)
        sys.exit(1)

    secret = os.environ.get("MATIMO_APPROVAL_SECRET")
    if not secret:
        print("❌ MATIMO_APPROVAL_SECRET is not set.", file=sys.stderr)
        print("   Set it to approve tools: export MATIMO_APPROVAL_SECRET=<your-secret>")
        sys.exit(1)

    yaml_path = Path(directory) / tool_name / "definition.yaml"
    if not yaml_path.is_file():
        print(f'❌ Cannot find definition.yaml for tool "{tool_name}" at:\n   {yaml_path}', file=sys.stderr)
        sys.exit(1)

    content = yaml_path.read_text(encoding="utf-8")

    # Promote status to "approved"
    try:
        parsed = yaml.safe_load(content) or {}
        if isinstance(parsed, dict) and parsed.get("status") != "approved":
            parsed["status"] = "approved"
            content = yaml.dump(parsed, default_flow_style=False)
            tmp_path = yaml_path.with_suffix(".yaml.tmp")
            tmp_path.write_text(content, encoding="utf-8")
            tmp_path.rename(yaml_path)
            print("   📝 Updated status: draft → approved in definition.yaml")
    except Exception:
        print("⚠️  Failed to update status in definition.yaml; proceeding with manifest approval only.")

    hash_val = manifest.compute_hash(content)
    approved_by = os.environ.get("USER") or os.environ.get("USERNAME") or "cli"
    manifest.approve(tool_name, hash_val, approved_by)
    print(f'✅ Tool "{tool_name}" approved.')


def _reject_tool(tool_name: str, directory: str) -> None:
    if not tool_name:
        print("❌ Usage: matimo review reject <tool-name>", file=sys.stderr)
        sys.exit(1)

    manifest = _try_load_manifest(directory)
    if manifest is None:
        print("❌ matimo core is not available.", file=sys.stderr)
        sys.exit(1)

    was_approved = manifest.revoke(tool_name)
    pending = manifest.get_pending_tools()
    was_pending = tool_name in pending

    if not was_approved and not was_pending:
        print(f'ℹ️  No record of tool "{tool_name}". Nothing to reject.')
        return

    print(f'🗑  Tool "{tool_name}" has been rejected/revoked.')
    if was_approved:
        print("   (Approval signature removed — the tool will be blocked until re-approved.)")


def review_command(args: list[str]) -> None:
    sub = args[0] if args else None
    directory = _resolve_manifest_dir()

    match sub:
        case "list" | None:
            _list_pending(directory)
        case "approve":
            _approve_tool(args[1] if len(args) > 1 else "", directory)
        case "reject":
            _reject_tool(args[1] if len(args) > 1 else "", directory)
        case _:
            print(f'❌ Unknown review subcommand: "{sub}"', file=sys.stderr)
            print("Usage: matimo review [list|approve|reject] [tool-name]")
            sys.exit(1)
